csi/csi.go

```diff
@@ -0,0 +1,158 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"github.com/red-hat-storage/ocs-client-operator/pkg/utils"
+	"github.com/red-hat-storage/ocs-client-operator/templates"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// +kubebuilder:rbac:groups="apps",resources=deployments,verbs=get;create;update;delete
+// +kubebuilder:rbac:groups="apps",resources=deployments/finalizers,verbs=update
+// +kubebuilder:rbac:groups="apps",resources=daemonsets,verbs=get;create;update;delete
+// +kubebuilder:rbac:groups="apps",resources=daemonsets/finalizers,verbs=update
+// +kubebuilder:rbac:groups="storage.k8s.io",resources=csidrivers,verbs=get;create;update;delete
+// +kubebuilder:rbac:groups="",resources=configmaps,verbs=create;update;delete
+// +kubebuilder:rbac:groups=security.openshift.io,resources=securitycontextconstraints,verbs=get;watch;create;patch;update
+// +kubebuilder:rbac:groups="",resources=serviceaccounts,verbs=get;create;update;delete
+// +kubebuilder:rbac:groups="",resources=serviceaccounts/finalizers,verbs=update
+// +kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=clusterroles,verbs=bind;escalate;get;create;update;delete
+// +kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=clusterroles/finalizers,verbs=update
+// +kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=clusterrolebindings,verbs=bind;escalate;get;create;update;delete
+// +kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=clusterrolebindings/finalizers,verbs=update
+// +kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=roles,verbs=bind;escalate;get;create;update;delete
+// +kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=roles/finalizers,verbs=update
+// +kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=rolebindings,verbs=bind;escalate;get;create;update;delete
+// +kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=rolebindings/finalizers,verbs=update
+
+// Proving escalate permission to operator as its required for creating
+// clusterrole and clusterrolebinding. More info:
+// https://kubernetes.io/docs/reference/access-authn-authz/rbac/#privilege-escalation-prevention-and-bootstrapping
+func CreateOrUpdateCSI(ctx context.Context, client client.Client, log logr.Logger) error {
+	operatorDeployment, err := getOperatorDeployment(ctx, client, utils.GetOperatorNamespace())
+	if err != nil {
+		log.Error(err, "unable to get operator deployment")
+		return err
+	}
+
+	err = createOrUpdateRBAC(ctx, client, operatorDeployment, log)
```

Comment:
For this and all the `create*` functions, would it make sense to rename them to `ensure*` since we should be ensuring they all exist and are up to date?